### ISSUE TITLE
[Discussion] Restructuring squidlib-util into single-purpose projects

### ISSUE THREAD

**Issue by tommyettinger** [yellowstonegames/SquidLib#142]
(Continuing from a note on #141 , I'd like to start a discussion about potential alternatives to the current "monolithic" squidlib-util project, ideally for after a stable 3.0.0 release.)

I think squidlib-util is getting too broad in scope and could or should be refactored into smaller, inter-dependent projects. This would allow library users to only include what they need and avoid adding unnecessary methods or expensive initialization steps. An example would be the roughly 256 KB (much more on GWT) of Hilbert curve data `squidpony.squidmath.CoordPacker` produces when it (or just `squidpony.squidmath.*`) is imported, even though the full range of Hilbert curve data is only used by very few parts of SquidLib. Another is the rather heavy-weight supply of more than 67,000 Coord values in the cache, which could be considered wasteful if an application isn't using Coord-like data much, or prefers to use another data structure for it (like LibGDX's Vector2). I've already copied some narrow parts of SquidLib into other stand-alone libraries, like [sarong](https://github.com/tommyettinger/sarong) for random numbers and hashing (the name is an acronym for Stand-Alone RandOm Number Generators), and [Gabber](https://github.com/tommyettinger/Gabber) for various text-generation things currently in SquidLib as FakeLanguageGen, NaturalLanguageCipher, and Messaging.

The data structures like OrderedMap employed throughout SquidLib are a bigger challenge, since a third-party application can't easily or efficiently mix SquidLib's OrderedMap with some other library's OrderedMap. There could be rather troubling dependency issues if GameFoo depends on SquidLib and LibBar, but LibBar doesn't depend on SquidLib and instead uses a narrow-purpose lib (let's call it LibBaz) that provides just data structures like OrderedMap, without compatibility between SquidLib's OrderedMap and LibBaz's OrderedMap. A solution for this last case is to make SquidLib, at least the -util parts of it, depend on a different library for data structures, and to allow other applications to depend on that library on its own without all of squidlib-util. There are some issues with this approach, one of which involves more work in gwt-inherits since each dependency's dependencies need to be at the same version (or possibly a compatible one) and each needs to be listed in the .gwt.xml file for the application. The data structure library would probably have a dependency on sarong for random numbers (and shuffling/reordering/fetching random elements from data structures) and the more obvious use of the hashing code in sarong, but might not need anything else.

General 2D grid code would be in a different library, including Coord, Zone classes like GreasedRegion, Radius, shape-related things like Rectangle, and possibly the various line-drawing classes like Bresenham. Dungeons would probably stay in squidlib-util, since the roguelike-and-related-game focus fits well for the main library.

I think squidlib, that is, the display code, would probably be untouched, depending on squidlib-util and because of that transitively depending on whatever squidlib-util depends on. Here we would benefit from what Maven and Gradle do so well, automatically downloading dependencies from Maven Central when you build.

For a concrete example, my toy language [celadon](https://github.com/tommyettinger/celadon) currently depends on SquidLib because it couldn't be used to script a game that uses SquidLib unless the data structures are the same, and though Celadon shouldn't actually depend on a game dev library, the possibility of a game depending on it forces my hand. If SquidLib had its various features pulled out into external libraries or just extra Maven modules in the SquidPony/SquidLib repo, then non-game projects could use what they need and game projects can then use those non-game projects while retaining compatibility with SquidLib.

I only see a few potential problems with this approach, but more substantial potential benefits. **Any thoughts on this?** Does anyone have experience with this approach or opinions on its effectiveness?

**Comment by smelc** [comment]
I think it'd indeed be nice if the project was split into more subprojects. I think SquidLib's core consists of the dungeon generating algorithms, FOV, LOS, DijkstraMap, AStarSearch, CoordPacker, GreasedRegion, Zone, ZOI, RNG, etc. More exotic features (FakeLanguageGen that I know of, and others I guess) would be better off in spinoff projects. For example, in my setup it would make sense not to depend on [RegExodus](https://github.com/tommyettinger/RegExodus), as I don't use code that depends on it.

**Comment by SquidPony** [comment]
I think splitting it apart more could be fairly challenging and dilute the usability of it for new users. That being said though, if the install tool can be made to work with all the parts in a friendly way and the library can be more useful to more people by splitting bits up more, I'm for it.

I take it as a good sign that there's so much meat that we have some growing pains on the structure of it :D

**Comment by tommyettinger** [comment]
@SquidPony , that's a good point re: usability, but if new users are't going through the source itself, then there's not much difference as long as the docs are clearly linked. One idea I'm considering near the top of the list is keeping everything in the same repo (to help keep links to SquidPony/SquidLib intact), but using Maven modules like the current split between `squidlib`, `squidlib-util`, `squidlib-extra`, and so on. There'd just be more modules, though they could be added as distinct dependencies. Building the parent project would rebuild docs for all subprojects, they'd all be in our /docs folder, and the versions would be kept 100% in sync for when a release is scheduled. It also keeps current contributors in the same system.

**Comment by tommyettinger** [comment]
@smelc , I'm not totally sure if RegExodus is limited to just the text-processing classes. There's a few cases where I need lists of upper-case letters, lower-case letters, all letters, and so on for certain uses (mostly where symbols need to be assigned to a potentially large range of items, and adding punctuation chars as symbols that represent one thing where some of those symbols already have meanings as other things would be a problem). A handful of cases are present elsewhere, and I think the display module uses RegExodus in one or two input-related classes to check if input is a letter, possibly to allow command words instead of just command keypresses, but I can't remember the specifics. It's a rather small lib; the code jar is 150 kb and the sources jar is 129 kb, plus GWT should remove a decent-sized chunk that's only present for compatibility with `java.util.regex` and isn't used by SquidLib.

That said, the parts SquidLib uses of RegExodus for non-regular-expression purposes (letter category data and such) could also be split off into a narrow-purpose library that RegExodus depends on and some parts of SquidLib might as well.

**Comment by smelc** [comment]
I agree that keeping one github repo would be nice as it makes clear what's SquidLib entry point is. Splitting the sources in different subfolder like we do for squidlib-util, squidlib-extra sounds good to me.

**Comment by SquidPony** [comment]
Also agreed on single entry with split packages. Jackson does that and
while a bit confusing it's at least easy enough to see what parts exist.

On Jan 26, 2017 00:47, "smelc" <notifications@github.com> wrote:

I agree that keeping one github repo would be nice as it makes clear what's
SquidLib entry point is. Splitting the sources in different subfolder like
we do for squidlib-util, squidlib-extra sounds good to me.

—
You are receiving this because you were mentioned.

Reply to this email directly, view it on GitHub
<https://github.com/SquidPony/SquidLib/issues/142#issuecomment-275335800>,
or mute the thread

**Comment by tommyettinger** [comment]
Glad there's a consensus. I'll probably start cleanup on unused classes in SquidLib's current setup next, then issue release 3.0.0-b7 shortly afterwards. After a round of hopefully-rather-intensive checks on b7 to ensure it's as bug-free as we can get it, that would be time for release 3.0.0 with the current split of squidlib-util, squidlib, and squidlib-extra. Beyond that, I think 3.x would become a branch and development would switch to 4.x with the deeper subdivision of the library into more modules. I'm not sure how well I can manage branches as a Collaborator (non-Owner) on GitHub, so I might need to coordinate with @SquidPony on the 4.x setup when we get there.

So for the parts that can probably be removed due to being unused or unusable...

WaffleFill is marked Beta and never exactly worked (nor would it ever be suitable for dungeon generation in games due to the lack of a guarantee that it can actually find a valid tiling), so it and the minor things that were introduced solely to facilitate it will be removed. Some classes come to mind in the mapping packages as being empty placeholders, so since they have had no content in them for at least a year, I think they can be removed... There's a few display classes that @smelc added early on in the 3.x cycle that he then improved upon later with other classes, and I'm not sure if removing the few early ones that are now marked Deprecated would be good to do now or later. I need to familiarize myself more with the code @smelc added to the display module, since a lot of the classes in there do similar things with small (and important) differences, and removing classes that seem to have duplicate purposes should only be done if a class has been made truly obsolete by another.

The Gradle build @smelc suggested in a recent PR might be worthwhile to pursue at some point, but I am not sure what advantages Gradle would give us over Maven, since Maven seems to work fine for now despite being older. I've really had an experience with Gradle that would be summarized with "mostly headaches, especially when IDEs are involved, and not worth the saved typing" compared to a Maven summary like "verbose, occasionally causes hair to be ripped out, but mostly works." Gradle also seems to be  slower than Maven, and the daemons it creates don't always get reused, leading to dozens of heavy daemons all running in the background unless someone on the building computer forcibly kills the processes. Since I've finally fixed a bug with Maven builds that could only happen on other developer's machines (GPG-related because releasing to Maven Central requires it, and it needed some external configuration to have my key specifically to build), people should have perfectly fine times building with Maven now (and only I can release to Central anyway, so GPG config should only be a requirement for me). Sorry that took so long for me to figure out!

**Comment by smelc** [comment]
I think all the classes marked deprecated in squidpony.squidgrid.gui.gdx should be deleted soon, as I don't use them anymore, and hence I won't be able to maintain them.

Please note that I use most of the classes there that I wrote that aren't deprecated, such as AbstractSquidScreen, LinesPanel, and SquidPanelBuilder. One exception is TextPanel which I forked in Dungeon Mercenary, because I was unhappy with a SquidLib change there and SquidApplicationAdapter which I never used, but is a good starting point for newcomers.

So to sum up, I would delete all deprecated classes there, plus TextPanel; because I won't maintain them.

Regarding gradle, I would leave my build file here (in other words: accept the PR ;-); it doesn't harm anything and works despite not being perfect; so it would help newcomers that use gradle.

**Comment by tommyettinger** [comment]
OK, I'll accept the PR as best as I can (there are merge conflicts), and refactor after that so it downloads dependencies from Maven Central instead of requiring a challenging build setup. I'll also remove the deprecated classes in the display module, and TextPanel.
